Add warning when using an empty regexp in lineinfile

samdoran opened the pull request:
##### SUMMARY

Reverts #41451 and adds a warning instead.

See #29443 for reasoning behind this change.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request


##### COMPONENT NAME
`lineinfile.py`

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
2.7
```

samdoran committed:
Use context managers for interacting with files

samdoran committed:
Store line and regexp parameters in a variable

samdoran committed:
Revert "Account for empty string regexp in lineinfile (#41451)"

This reverts commit 4b5b4a760c5217bdfd1d45efd224e2762e9898b0.

samdoran committed:
Add warning when regexp is an empty string

abadger commented on lib/ansible/modules/files/lineinfile.py:
```diff
@@ -480,12 +480,17 @@ def main():
 
+    if regexp == '':
+        module.warn(
+            'The regular expression is empty, which will match every line in the file.'
```
I think we need to say "empty string".  We also may want to give an alternative for people who actually do want to match every line in the file.

samdoran replied:
What do you mean by "give an alternative"? This warning could be annoying to someone who does want this capability, and I don't want to encourage folks to disable warnings with noisy error messages.

abadger replied:
Give an example of a regex to use if they really want to match every line.  (I think "^"  would do it and be efficient).  We could also give an example of unsetting it but that isn't as succinctly written.

samdoran replied:
Ah, that's perfect! I see what you're saying.

abadger commented on test/integration/targets/lineinfile/tasks/main.yml:
```diff
@@ -717,37 +717,3 @@
   assert:
     that:
       - result.stat.checksum == 'eca8d8ea089d4ea57a3b87d4091599ca8b60dfd2'
```
Keeping tests (but updating if the results have changed) seems like a good idea.  Then it's an additional clue that the behaviour was meant to be this way.

samdoran committed:
Remove '=' from error messages

abadger commented:
Also, I think we want a changelog for this?  (And maybe a porting guide?)

samdoran committed:
Update warning message and add changelog

samdoran committed:
Add tests

samdoran committed:
Improve warning message

Offer an equivalent regexp that won't trigger the warning.
Update tests to match new warning.

samdoran committed:
Add porting guide entry for lineinfile change
